src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -456,15 +476,15 @@ protected void AbortRequest()
         {
             lock (_abortLock)
             {
-                if (_requestAborted)
+                if (_connectionAborted)
                 {
                     return;
                 }
 
-                _requestAborted = true;
+                _connectionAborted = true;
             }
 
-            if (_abortedCts != null)
+            if (_abortedCts != null && !_preventRequestAbortedCancellation)
```

Comment:
_preventRequestAbortedCancellation and _abortedCts need to be checked under the lock?

Reply:
I agree. I was about to say the real _preventRequestAbortedCancellation and _abortedCts checks happen inside the scheduled CancelRequestAbortedToken method, but _preventRequestAbortedCancellation is resettable, so I'd feel better if we determined a `shouldScheduleCancellation` condition inside the above _abortLock.

Reply:
I found [the comment](https://github.com/aspnet/KestrelHttpServer/pull/2977#discussion_r222387365) I made to @Tratcher the last time he brought this up. I'm not sure I was wrong, but I've learned since then to be extremely careful with this code.

Reply:
I'll go ahead and do the local.

Reply:
While you're at it move the `_preventRequestAbortedCancellation = false` in Reset() down a few lines into the _abortLock.

Like the local, I don't think it's truly necessary, but I think Reset() is the last place _preventRequestAbortedCancellation is accessed without a lock, and I'd feel feel warm and fuzzy if all access was done with a lock.